OrbitGl/TimerChain.h

```diff
@@ -0,0 +1,86 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_GL_TIMER_CHAIN_
+#define ORBIT_GL_TIMER_CHAIN_
+
+#include <assert.h>
+
+#include <algorithm>
+#include <atomic>
+#include <cstdint>
+#include <limits>
+
+#include "TextBox.h"
+
+class TimerBlock {
+ public:
+  TimerBlock();
+
+  bool AtCapacity() const {
+    return size_ == kMaxSize;
+  }
+
+  // Silently fails when the block is full.
+  void Add(const TextBox& box);
+  
+  bool Intersects(uint64_t min, uint64_t max);
+
+  int size() const {
+    return size_;
+  }
+
+  uint64_t min_timestamp() {
+    return min_timestamp_;
+  }
+
+  uint64_t max_timestamp() {
+    return max_timestamp_;
+  }
+
+  TextBox& operator[](std::size_t idx) { 
+    return data_[idx];
+  }
+    
+  const TextBox& operator[](std::size_t idx) const { 
+    return data_[idx];
+  }
+  
+ private:
+  static constexpr int kMaxSize = 1024;
+  TextBox data_[kMaxSize];
+  std::atomic<int> size_;
+
+  uint64_t min_timestamp_;
+  uint64_t max_timestamp_;
+};
+
+class TimerChain {
+ public:
+  TimerChain();
+  ~TimerChain();
+
+  void push_back(const TextBox& box);
+
+  int size() const { return blocks_.size(); }
+
+  TimerBlock* operator[](std::size_t idx) { 
+    return blocks_[idx];
+  }
+    
+  const TimerBlock* operator[](std::size_t idx) const { 
+    return blocks_[idx];
+  }
+
+  int GetBlockContaining(const TextBox* element);
+
+  TextBox* GetElementAfter(const TextBox* element);
+
+  TextBox* GetElementBefore(const TextBox* element);
+
+ private:
+  std::vector<TimerBlock*> blocks_;
```

Comment:
I believe the way BlockChain / TimerChain is used, vector was a buggy choice. The data is written by a single thread and read from different threads, so any reallocation of the vector data will lead to obvious crashes (and would be inefficient anyway; we also don't seem to need random access anyway). I'm not yet sure the previous implementation was entirely bug free regarding accesses from multiple threads, so maybe we also need to introduce some locking.

Reply:
The initial implementation was a rudimentary lockless single-writer-multiple-readers data structure.  Adding an element to a block atomically increments its size (note that std::atomic was removed recently as it was not needed on x64 for a 64 bit (aligned) access).  When a block is full, a new one is allocated and the "m_Next" pointer of the previous block is (atomically) assigned.  The idea was to have timers coalesced in memory for cache friendliness and the lockless aspect was ideal because when data is streaming in, there is a lot of contention, especially with a not-so-optimized rendering loop. 

I would be curious to compare the performance to having a lock and a map (having a binary tree would allow us to quickly find elements that need to be drawn). Another approach would be to enqueue items to be appended to the "chain" and add them in bulk on the main thread.  The chain would then be accessed only from a single thread.  This pattern is used in the codebase using the LockFreeQueue data structure to pass data between threads.

Reply:
By the way, the BlockChain data structure was never properly tested and I'm all for moving away from it.

Reply:
Cool, thanks for the details. I now went with a straightforward specialization of BlockChain with added min/max timestamps. I agree with need to look into this in more detail at a later time, though.